File: scripts/fetch_images.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Télécharge des photos d'ambiance depuis l'API officielle Unsplash pour l'affiche
de l'APE, dans assets/, et enregistre les crédits obligatoires dans
assets/credits.txt (photographe + liens, conditions Unsplash).

- Lit la clé UNSPLASH_ACCESS_KEY depuis le fichier .env à la racine.
- N'utilise que la bibliothèque standard (aucun `pip install` nécessaire).
- Respecte les guidelines Unsplash : authentification Client-ID, déclenchement
  de l'endpoint `download_location`, et attribution avec UTM.

Usage :
    python scripts/fetch_images.py

Une fois les images téléchargées, régénérer le PDF :
    node scripts/generate-pdf.js src/affiche-a3.html dist/affiche-ape-a3.pdf
"""

import json
import os
import sys
import urllib.parse
import urllib.request

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ASSETS = os.path.join(ROOT, "assets")
ENV_PATH = os.path.join(ROOT, ".env")

APP_NAME = "ape-segoufielle"  # pour les paramètres UTM d'attribution
API = "https://api.unsplash.com"

# Mots-clés d'ambiance (on évite les portraits d'enfants identifiables).
# slug -> requête Unsplash
KEYWORDS = {
    "children-craft-hands": "children craft hands",
    "colorful-balloons-party": "colorful balloons party",
    "school-fair-festival": "school fair festival",
    "kids-painting-art-supplies": "kids painting art supplies",
    "community-celebration-outdoor": "community celebration outdoor",
}


def load_access_key():
    """Lit UNSPLASH_ACCESS_KEY depuis .env (parseur minimal)."""
    if not os.path.exists(ENV_PATH):
        sys.exit("Erreur : fichier .env introuvable à la racine du projet.")
    key = None
    with open(ENV_PATH, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            name, _, value = line.partition("=")
            if name.strip() == "UNSPLASH_ACCESS_KEY":
                key = value.strip().strip('"').strip("'")
    if not key:
        sys.exit(
            "Erreur : UNSPLASH_ACCESS_KEY est vide dans .env.\n"
            "Crée une application sur https://unsplash.com/oauth/applications\n"
            "puis colle l'Access Key dans .env."
        )
    return key


def api_get(url, access_key):
    req = urllib.request.Request(
        url,
        headers={
            "Authorization": "Client-ID " + access_key,
            "Accept-Version": "v1",
            "User-Agent": APP_NAME,
        },
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read().decode("utf-8"))


def search_photo(query, access_key):
    """Renvoie le 1er résultat paysage pour la requête, ou None."""
    params = urllib.parse.urlencode(
        {
            "query": query,
            "orientation": "landscape",
            "per_page": 5,
            "content_filter": "high",
        }
    )
    data = api_get(API + "/search/photos?" + params, access_key)
    results = data.get("results", [])
    return results[0] if results else None


def trigger_download(photo, access_key):
    """Guideline Unsplash : pinger download_location quand on utilise une photo."""
    loc = photo.get("links", {}).get("download_location")
    if loc:
        try:
            api_get(loc, access_key)
        except Exception as e:  # non bloquant
            print("  (avertissement) download_location :", e)


def download_image(photo, dest):
    """Télécharge une version adaptée à l'impression (largeur ~2000px, jpg)."""
    raw = photo["urls"]["raw"]
    sep = "&" if "?" in raw else "?"
    url = raw + sep + urllib.parse.urlencode(
        {"w": 2000, "q": 80, "fm": "jpg", "fit": "crop"}
    )
    req = urllib.request.Request(url, headers={"User-Agent": APP_NAME})
    with urllib.request.urlopen(req, timeout=60) as resp, open(dest, "wb") as out:
        out.write(resp.read())


def utm(url):
    sep = "&" if "?" in url else "?"
    return url + sep + urllib.parse.urlencode(
        {"utm_source": APP_NAME, "utm_medium": "referral"}
    )


def main():
    access_key = load_access_key()
    os.makedirs(ASSETS, exist_ok=True)

    credits = [
        "Crédits photos — Unsplash (https://unsplash.com)",
        "Attribution obligatoire selon les conditions Unsplash.",
        "=" * 60,
        "",
    ]

    for slug, query in KEYWORDS.items():
        print(f"Recherche « {query} »…")
        photo = search_photo(query, access_key)
        if not photo:
            print("  Aucun résultat, ignoré.")
            continue

        dest = os.path.join(ASSETS, slug + ".jpg")
        trigger_download(photo, access_key)
        download_image(photo, dest)

        user = photo.get("user", {})
        name = user.get("name", "Inconnu")
        profile = utm(user.get("links", {}).get("html", "https://unsplash.com"))
        photo_link = utm(photo.get("links", {}).get("html", "https://unsplash.com"))
        print(f"  -> {slug}.jpg  (Photo de {name})")

        credits.append(f"{slug}.jpg")
        credits.append(f"  Photo de {name} sur Unsplash")
        credits.append(f"  Profil : {profile}")
        credits.append(f"  Photo  : {photo_link}")
        credits.append("")

    with open(os.path.join(ASSETS, "credits.txt"), "w", encoding="utf-8") as f:
        f.write("\n".join(credits))

    print("\nTerminé. Crédits écrits dans assets/credits.txt")
    print("Régénère le PDF : node scripts/generate-pdf.js src/affiche-a3.html dist/affiche-ape-a3.pdf")


if __name__ == "__main__":
    main()

File: scripts/fetch-images.js
// Équivalent Node de scripts/fetch_images.py (utile si Python n'est pas installé).
// Télécharge des photos d'ambiance via l'API officielle Unsplash dans assets/,
// et écrit les crédits obligatoires (photographe + liens) dans assets/credits.txt.
// Lit la clé depuis .env. Respecte les guidelines Unsplash (Client-ID,
// ping download_location, attribution UTM).
//
// NB : on passe par `curl` (child_process) car `fetch` de Node échoue dans
// certains environnements (proxy/TLS) où curl fonctionne.
//
// Usage : node scripts/fetch-images.js

const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

const ROOT = path.join(__dirname, '..');
const ASSETS = path.join(ROOT, 'assets');
const ENV_PATH = path.join(ROOT, '.env');
const APP = 'ape-segoufielle';
const API = 'https://api.unsplash.com';
const NULLDEV = process.platform === 'win32' ? 'NUL' : '/dev/null';

const KEYWORDS = {
  'children-craft-hands': 'children craft hands',
  'colorful-balloons-party': 'colorful balloons party',
  'school-fair-festival': 'school fair festival',
  'kids-painting-art-supplies': 'kids painting art supplies',
  'community-celebration-outdoor': 'community celebration outdoor',
};

function loadKey() {
  if (!fs.existsSync(ENV_PATH)) throw new Error('.env introuvable');
  for (const line of fs.readFileSync(ENV_PATH, 'utf8').split(/\r?\n/)) {
    const s = line.trim();
    if (!s || s.startsWith('#') || !s.includes('=')) continue;
    const [name, ...rest] = s.split('=');
    if (name.trim() === 'UNSPLASH_ACCESS_KEY') {
      return rest.join('=').trim().replace(/^['"]|['"]$/g, '');
    }
  }
  throw new Error('UNSPLASH_ACCESS_KEY absent de .env');
}

const AUTH = (key) => ['-H', `Authorization: Client-ID ${key}`, '-H', 'Accept-Version: v1'];

function curlJson(url, key) {
  const out = execFileSync('curl', ['-s', ...AUTH(key), url], {
    encoding: 'utf8',
    maxBuffer: 32 * 1024 * 1024,
  });
  return JSON.parse(out);
}

function curlPing(url, key) {
  try { execFileSync('curl', ['-s', '-o', NULLDEV, ...AUTH(key), url]); } catch {}
}

function curlDownload(url, dest) {
  execFileSync('curl', ['-s', '-L', '-o', dest, '-A', APP, url], {
    maxBuffer: 64 * 1024 * 1024,
  });
}

const utm = (u) =>
  u + (u.includes('?') ? '&' : '?') + `utm_source=${APP}&utm_medium=referral`;

function main() {
  const key = loadKey();
  fs.mkdirSync(ASSETS, { recursive: true });

  const credits = [
    'Crédits photos — Unsplash (https://unsplash.com)',
    'Attribution obligatoire selon les conditions Unsplash.',
    '='.repeat(60),
    '',
  ];

  for (const [slug, query] of Object.entries(KEYWORDS)) {
    process.stdout.write(`Recherche « ${query} »… `);
    const url =
      `${API}/search/photos?query=${encodeURIComponent(query)}` +
      `&orientation=landscape&per_page=5&content_filter=high`;
    const data = curlJson(url, key);
    const photo = (data.results || [])[0];
    if (!photo) { console.log('aucun résultat'); continue; }

    const dl = photo.links && photo.links.download_location;
    if (dl) curlPing(dl, key);

    const raw = photo.urls.raw;
    const imgUrl = raw + (raw.includes('?') ? '&' : '?') + 'w=2000&q=80&fm=jpg&fit=crop';
    curlDownload(imgUrl, path.join(ASSETS, slug + '.jpg'));

    const name = (photo.user && photo.user.name) || 'Inconnu';
    const profile = utm((photo.user && photo.user.links && photo.user.links.html) || 'https://unsplash.com');
    const link = utm((photo.links && photo.links.html) || 'https://unsplash.com');
    console.log(`-> ${slug}.jpg (Photo de ${name})`);

    credits.push(`${slug}.jpg`);
    credits.push(`  Photo de ${name} sur Unsplash`);
    credits.push(`  Profil : ${profile}`);
    credits.push(`  Photo  : ${link}`);
    credits.push('');
  }

  fs.writeFileSync(path.join(ASSETS, 'credits.txt'), credits.join('\n'), 'utf8');
  console.log('\nTerminé. Crédits dans assets/credits.txt');
}

main();
